oak/server/rust/oak_runtime/src/channel.rs

```diff
@@ -126,7 +136,18 @@ impl ChannelWriter {
         if self.is_orphan() {
             return Err(OakStatus::ERR_CHANNEL_CLOSED);
         }
+
         let mut messages = self.messages.write().unwrap();
+        let mut waiting_threads = self.waiting_threads.lock().unwrap();
+
+        // Unpark all the waiting threads, that still have references
+        for thread in waiting_threads.values() {
+            if let Some(thread) = thread.upgrade() {
+                thread.unpark();
```

Comment:
Isn't this racy?  If the other thread immediately starts up and tries to read from the channel, it won't find a message (l.151 hasn't happened yet), and will then re-park forever.

Reply:
Upon being woken up there is no guarantee a message is available. The lock on `messages` above prevents any woken threads from reading `messages` before a message is added below. All threads are guaranteed to be woken before they can observe `messages`, and so all get a chance to read the message. 

However, after the change above this is slightly different- Now the first channel woken can immediately read the message, and others might find `messages` is empty before they are even woken. This should not be an issue (being woken does not guarantee a message is available, and there is no race here on waiting again), but it could potentially result in some particular thread always getting first chance to read the message. The previous code would likely be biased to the first channel woken too.

So there is some bias in which thread is most likely to receive the message, but this shouldn't be a problem.

Reply:
Aha, not racy &ndash; but it was kind of relying on the locking hierarchy (messages ⊲ waiting_threads) that got removed above!

Reply:
I think it would be nice to add some of these comments to the code, I found them quite useful while trying to understand the code, and it would have taken me much longer without them.